File: backend/app/api/routes/quick_practice.py
"""Student-facing multi-machine quick-practice API."""

import uuid

from fastapi import APIRouter

from app.api.deps import AdminUser, CurrentUser, SessionDep
from app.services import quick_practice as quick_practice_service
from app.services.scheduling.recurrence import get_schedule_policy

router = APIRouter(prefix="/quick-practice", tags=["quick-practice"])


def _serialize_template(session: SessionDep, environment, version) -> dict:
    nodes = quick_practice_service.nodes_for_version(
        session, version_id=version.id
    )
    return {
        "id": environment.id,
        "version_id": version.id,
        "name": environment.name,
        "description": environment.description,
        "version": version.version,
        "status": version.status,
        "usage_scope": environment.usage_scope,
        "published_at": version.published_at,
        "duration_hours": get_schedule_policy(session=session).practice_session_hours,
        "nodes": [node.model_dump() for node in nodes],
        "per_student": {
            "machines": len(nodes),
            "cpu_cores": sum(node.cpu for node in nodes),
            "memory_mb": sum(node.memory_mb for node in nodes),
            "disk_gb": sum(node.disk_gb for node in nodes),
        },
    }


@router.get("/templates")
def list_templates(
    session: SessionDep, current_user: CurrentUser
) -> list[dict]:
    return [
        _serialize_template(session, environment, version)
        for environment, version in quick_practice_service.list_published_templates(
            session, user=current_user
        )
    ]


@router.get("/templates/{environment_id}")
def get_template(
    environment_id: uuid.UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> dict:
    environment, version = quick_practice_service.get_published_template(
        session, environment_id=environment_id, user=current_user
    )
    return _serialize_template(session, environment, version)


@router.post("/templates/{environment_id}/launch", status_code=202)
def launch_template(
    environment_id: uuid.UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> dict:
    item = quick_practice_service.launch(
        session, user=current_user, environment_id=environment_id
    )
    return quick_practice_service.serialize_session(session, item)


@router.post("/sessions/{practice_id}/end")
def end_session(
    practice_id: uuid.UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> dict:
    """提早結束自己的練習，整組立刻進入回收。"""
    item = quick_practice_service.end_session(
        session, user=current_user, practice_id=practice_id
    )
    return quick_practice_service.serialize_session(session, item)


@router.get("/sessions/my")
def list_my_sessions(
    session: SessionDep, current_user: CurrentUser
) -> list[dict]:
    return [
        quick_practice_service.serialize_session(session, item)
        for item in quick_practice_service.list_sessions(
            session, user_id=current_user.id
        )
    ]


@router.get("/sessions")
def list_all_sessions(
    session: SessionDep, _current_user: AdminUser
) -> list[dict]:
    return [
        quick_practice_service.serialize_session(session, item)
        for item in quick_practice_service.list_sessions(session)
    ]

File: frontend/src/services/quickPractice.js
import { apiGet, apiPost } from "./api";

function normalizeNode(node) {
  return {
    ...node,
    id: node.node_key ?? String(node.id),
    type: node.resource_type,
    memory: Math.max(1, Math.round(Number(node.memory_mb ?? 1024) / 1024)),
    disk: Number(node.disk_gb ?? 0),
  };
}

function normalizeTemplate(item) {
  return {
    ...item,
    id: String(item.id),
    versionId: String(item.version_id),
    usageScope: item.usage_scope,
    nodes: (item.nodes ?? []).map(normalizeNode),
  };
}

function normalizeSession(item) {
  return {
    ...item,
    id: String(item.id),
    kindLabel: item.kind_label,
    environmentId: String(item.environment_id),
    environmentVersionId: String(item.environment_version_id),
    expiresAt: item.expires_at,
    createdAt: item.created_at,
    machines: (item.machines ?? []).map((machine) => ({
      ...machine,
      id: String(machine.id),
      requestId: String(machine.request_id),
      type: machine.resource_type,
      ip: machine.ip_address,
    })),
  };
}

export const QuickPracticeService = {
  async listTemplates(options) {
    return (await apiGet("/api/v1/quick-practice/templates", options)).map(normalizeTemplate);
  },
  async getTemplate(environmentId) {
    return normalizeTemplate(await apiGet(`/api/v1/quick-practice/templates/${environmentId}`));
  },
  async launch(environmentId) {
    return normalizeSession(await apiPost(`/api/v1/quick-practice/templates/${environmentId}/launch`, {}));
  },
  async endSession(sessionId) {
    return normalizeSession(await apiPost(`/api/v1/quick-practice/sessions/${sessionId}/end`, {}));
  },
  async listMySessions(options) {
    return (await apiGet("/api/v1/quick-practice/sessions/my", options)).map(normalizeSession);
  },
  async listAllSessions(options) {
    return (await apiGet("/api/v1/quick-practice/sessions", options)).map(normalizeSession);
  },
};
